{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "#find the amount of cpg pairs there are within a string of 150 bases in the\n",
        "#sequence after methylation and return true if the sequence has at least 6 cpg\n",
        "#pairs\n",
        "#The input will be a sequence and the start of the section you are looking at\n",
        "#for potential primers\n",
        "def dense(newseq,start,first):\n",
        "  #print(start)\n",
        "  cgdense = 0\n",
        "  p = -1\n",
        "  if (start + 150 >= len(newseq)):\n",
        "    return -1\n",
        "  for i in range (start,start + 150):\n",
        "    if ((newseq[i] == 'C' or newseq[i] == 'c') and i != 149):\n",
        "      cgdense += 2;\n",
        "  if (first and cgdense >= 30):\n",
        "    return 0\n",
        "  else:\n",
        "    for i in range (start,len(newseq) - 150):\n",
        "      #print(i,cgdense)\n",
        "      if ((newseq[i] == 'C' or newseq[i] == 'c')):\n",
        "        cgdense -= 2\n",
        "      if ((newseq[i + 150] == 'C' or newseq[i + 150] == 'c')):\n",
        "        cgdense += 2\n",
        "      if (cgdense >= 12):\n",
        "        #print(newseq[i:i + 150])\n",
        "        #for a in range (150):\n",
        "        #  fout.write(newseq[i + a + 1])\n",
        "        p = i\n",
        "        break\n",
        "  if (p + 150 >= len(newseq) ):\n",
        "    return 0-1\n",
        "  return p\n",
        "#the function will return the first place in the sequence after the start\n",
        "#it finds has enough cpg pairs to serve as an amplicon"
      ],
      "metadata": {
        "id": "9iNkeishEGHZ"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#counts and returns the number of c and g in the given sequence\n",
        "#ACAGTCAGCTACGATCAGCATCAGCT will return 13\n",
        "def gcdensity(seq):\n",
        "  gc = 0\n",
        "  #for i in range (len(seq) - 1):\n",
        "  #  if (seq[i:i + 2] == \"GC\" or seq[i:i + 2] == \"gc\"):\n",
        "  #    gc += 1;\n",
        "  for i in range (len(seq)):\n",
        "    if (seq[i] == 'G' or seq[i] == 'g' or seq[i] == 'C' or seq[i] == 'c'):\n",
        "      gc += 1\n",
        "  return gc"
      ],
      "metadata": {
        "id": "kIrbDjKCmfLX"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Finds all pairs of two cgs in a sequence and considers\n",
        "#every combination of forward and reverse primer\n",
        "#that could be formed in the sequence and scores them by\n",
        "#how likely they would work in an assay before returning\n",
        "\n",
        "#The input is a sequence of 150 bases that could be a possible amplicon\n",
        "#The output is a list of possible amplicons\n",
        "#in the given sequence with scores to indicate\n",
        "#how likely they will contain a forward and reverse primer\n",
        "def cgcgpairs(newseq):\n",
        "  points = [];\n",
        "  #record the location of every cpg\n",
        "  for i in range (len(newseq)):\n",
        "    if (newseq[i] == 'C' or newseq[i] == 'c'):\n",
        "      points.append(i)\n",
        "      #print(i)\n",
        "  start = -1\n",
        "  at = 0\n",
        "  pairs = []\n",
        "  #records the location of pairs of cpgs\n",
        "  for i in range(len(points) - 1):\n",
        "    #print(points[i + 1] - points[i])\n",
        "    if (points[i + 1] - points[i] <= 3) :\n",
        "      pairs.append([points[i],points[i + 1]])\n",
        "      #start = points[i]\n",
        "      #at = i\n",
        "      #break\n",
        "  #end = -1\n",
        "  #for i in range(len(points) - 1,at,-1):\n",
        "    #print(points[i] - points[i - 1])\n",
        "  #  if (points[i] - points[i - 1] <= 3) :\n",
        "  #    end = points[i]\n",
        "  #    break\n",
        "  #if(end - start >= 100):\n",
        "  #  return \"-1\"\n",
        "\n",
        "  cgcgstr = \"\"\n",
        "  #cgcgstr += str(start) + \" \" + str(end)\n",
        "  if (len(pairs) == 1):\n",
        "    return \"-1\"\n",
        "  min = 100\n",
        "  global score\n",
        "  score = 0\n",
        "  f = -1\n",
        "  s = -1\n",
        "  #for i in range(len(pairs)):\n",
        "    #print(pairs[i][0])\n",
        "  for i in range(len(pairs) - 1):\n",
        "    #filters out primers with a cg pair too close to the start or end\n",
        "    if (pairs[i][0] < 21):\n",
        "      continue\n",
        "    #print(pairs[i][0])\n",
        "    for a in range(i + 1,len(pairs) - 1):\n",
        "      if (pairs[a][1] + 1 >= len(newseq) - 20):\n",
        "        continue\n",
        "      #give scores based on how many bases are between the first and last pair of cpgs\n",
        "      #used to punish cases of consecutive cpgs\n",
        "      curscore = 0\n",
        "      curscore += 5 + pairs[i][1] - pairs[i][0]\n",
        "      curscore += 5 + pairs[a][1] - pairs[a][0]\n",
        "      #make sure the distance between the first and last pair of cpgs is in the range of 40-100 bases\n",
        "      #used to make sure forward and reverse primer are placed at an appropriate range\n",
        "      if (pairs[a][1] - pairs[i][0] < 100 and pairs[a][0] - (pairs[i][1] + 1) > 40):\n",
        "        sum = pairs[a][1] - pairs[i][0]\n",
        "        #give scores to the primers based on how many bases they are seperated by score decreases by 1 every 20\n",
        "        curscore += 6 - (sum -(sum % 20)) / 20\n",
        "        if (curscore > score):\n",
        "          #the newseq cannot be composed of more than 75% c or g\n",
        "          if (gcdensity(newseq) > 3*len(newseq) / 4):\n",
        "            continue\n",
        "          #if (gcdensity(newseq[0:pairs[i][0] + 1]) > 3 * (pairs[i][0] + 1) / 4):\n",
        "          #  continue\n",
        "          #if (gcdensity(newseq[pairs[i][0]:pairs[a][1] + 2]) > 3 * len(newseq) / 4):\n",
        "          #  continue\n",
        "          #if (gcdensity(newseq[pairs[a][1] + 2:150]) > 3 * (150 - (pairs[a][1] + 2)) / 4):\n",
        "          #  continue\n",
        "          score = curscore\n",
        "          min = pairs[a][1] - pairs[i][0]\n",
        "          f = i\n",
        "          s = a\n",
        "          continue\n",
        "        #only takes the highest scoring pair for an amplicon\n",
        "        if (curscore < score):\n",
        "          continue\n",
        "        #every test the distance between the first and last pair is less\n",
        "        if (pairs[a][1] - pairs[i][0] > min):\n",
        "          break\n",
        "        #print(pairs[a][1] - pairs[i][0])\n",
        "        min = pairs[a][1] - pairs[i][0]\n",
        "        f = i\n",
        "        s = a\n",
        "  #print(pairs[f][0] + 1,pairs[s][0] + 1)\n",
        "  #print(f,s)\n",
        "  #print(f,s)\n",
        "  if (f == -1 or s == -1):\n",
        "    return \"-1\"\n",
        "  #print(newseq[pairs[f][0]:pairs[s][1] + 2])\n",
        "  cgcgstr = str(pairs[f][1] + 1) + \" \" + str(pairs[s][0] + 2)\n",
        "  #print(cgcgstr)\n",
        "  #if (gcdensity(newseq[pairs[f][0]:pairs[s][1] + 2]) > 7 * len(newseq) / 10):\n",
        "    #return \"-1\"\n",
        "  #record information about the location of the section in the whole strand\n",
        "  global splace\n",
        "  splace = pairs[f][0]\n",
        "  global eplace\n",
        "  eplace = pairs[s][1]\n",
        "  global splace2\n",
        "  splace2 = pairs[f][1]\n",
        "  global eplace2\n",
        "  eplace2 = pairs[s][0]\n",
        "  #print(newseq[start:end + 2])\n",
        "  return cgcgstr\n"
      ],
      "metadata": {
        "id": "WSU9Y4-_Hy5_"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#when given a strand returns the sequence of the complementary strand\n",
        "def complementary(seq):\n",
        "  newseq = \"\"\n",
        "  for i in range (len(seq)-1,-1,-1):\n",
        "    if (seq[i] == 'G' or seq[i] == 'g'):\n",
        "      newseq += 'C'\n",
        "    if (seq[i] == 'C' or seq[i] == 'c'):\n",
        "      newseq += 'G'\n",
        "    if (seq[i] == 'T' or seq[i] == 't'):\n",
        "      newseq += 'A'\n",
        "    if (seq[i] == 'A' or seq[i] == 'a'):\n",
        "      newseq += 'T'\n",
        "  return newseq"
      ],
      "metadata": {
        "id": "dWimw1Tzlcss"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#when given a strand returns the sequence of the complementary strand in reverse\n",
        "def reverse(seq):\n",
        "  newseq = \"\"\n",
        "  for i in range (len(seq)):\n",
        "    if (seq[i] == 'G' or seq[i] == 'g'):\n",
        "      newseq += 'C'\n",
        "    if (seq[i] == 'C' or seq[i] == 'c'):\n",
        "      newseq += 'G'\n",
        "    if (seq[i] == 'T' or seq[i] == 't'):\n",
        "      newseq += 'A'\n",
        "    if (seq[i] == 'A' or seq[i] == 'a'):\n",
        "      newseq += 'T'\n",
        "  return newseq"
      ],
      "metadata": {
        "id": "FEP2YZacBVlx"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#flip a strand for reverse primers\n",
        "def flip(seq):\n",
        "  newseq = \"\"\n",
        "  for i in range (len(seq)-1,-1,-1):\n",
        "    if (seq[i] == 'G' or seq[i] == 'g'):\n",
        "      newseq += 'G'\n",
        "    if (seq[i] == 'C' or seq[i] == 'c'):\n",
        "      newseq += 'C'\n",
        "    if (seq[i] == 'T' or seq[i] == 't'):\n",
        "      newseq += 'T'\n",
        "    if (seq[i] == 'A' or seq[i] == 'a'):\n",
        "      newseq += 'A'\n",
        "  return newseq"
      ],
      "metadata": {
        "id": "n2mAIxBpAyC_"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#finds if two strands have primer dimers between each other\n",
        "\n",
        "#input is two strands and an integer to indicate whether each strand is forward or reverse\n",
        "def diff(forw,rev,fr):\n",
        "  #print(forw,rev)\n",
        "  #rep = complementary(seq)\n",
        "  #print(forw)\n",
        "  #print(rev)\n",
        "  dimers = []\n",
        "  for i in range (2,min(len(forw),len(rev))):\n",
        "    for a in range (len(forw) - i + 1):\n",
        "      for e in range (len(rev) - i + 1):\n",
        "        #print (seq[a:i + a])\n",
        "        if (forw[a:i + a] == rev[e:i + e]):\n",
        "          if (fr != 3):\n",
        "            dimers.append([forw[a:i + a],len(forw) - a - i,e])\n",
        "          else:\n",
        "            dimers.append([forw[a:i + a],a,len(rev) - e - i])\n",
        "          #print(forw[a:i + a],len(forw) - a - i,e)\n",
        "          #print(len(forw) - a - i,e)\n",
        "  bad = False\n",
        "  for i in range(len(dimers) - 1,-1,-1):\n",
        "    if (dimers[i][1] != 0 and dimers[i][2] != 0):\n",
        "      continue\n",
        "    if (fr == 1 and (dimers[i][1] == 0 or dimers[i][2] == 0)):\n",
        "      if (len(dimers[i][0]) <= 2 or dimers[i][1] > 2 or dimers[i][2] > 2):\n",
        "        continue\n",
        "      print(forw)\n",
        "      spaces = \"\"\n",
        "      for a in range (len(forw) - len(dimers[i][0])):\n",
        "        spaces += \" \"\n",
        "      if (dimers[i][1] > 0):\n",
        "        spaces = spaces[0:len(spaces) - dimers[i][1]]\n",
        "      connect = spaces\n",
        "      for a in range (len(dimers[i][0])):\n",
        "        connect += \"|\"\n",
        "      if (dimers[i][2] > 0):\n",
        "        spaces = spaces[0:len(spaces) - dimers[i][2]]\n",
        "      print(connect)\n",
        "      print(spaces + reverse(rev))\n",
        "      bad = True\n",
        "    elif (fr == 2 and (dimers[i][1] == 0 or dimers[i][2] == 0)):\n",
        "      #print(dimers[i][0],dimers[i][1],dimers[i][2])\n",
        "      if (len(dimers[i][0]) <= 2 or dimers[i][1] > 2 or dimers[i][2] > 2):\n",
        "        continue\n",
        "      print(forw)\n",
        "      spaces = \"\"\n",
        "      for a in range (len(forw) - len(dimers[i][0])):\n",
        "        spaces += \" \"\n",
        "      if (dimers[i][1] > 0):\n",
        "        spaces = spaces[0:len(spaces) - dimers[i][1]]\n",
        "      connect = spaces\n",
        "      for a in range (len(dimers[i][0])):\n",
        "        connect += \"|\"\n",
        "      if (dimers[i][2] > 0):\n",
        "        spaces = spaces[0:len(spaces) - dimers[i][2]]\n",
        "      print(connect)\n",
        "      print(spaces + reverse(rev))\n",
        "      bad = True\n",
        "    elif (fr == 3 and (dimers[i][1] == 0 or dimers[i][2] == 0)):\n",
        "      #print(dimers[i][0],dimers[i][1],dimers[i][2])\n",
        "      if (len(dimers[i][0]) <= 2 or dimers[i][1] > 2 or dimers[i][2] > 2):\n",
        "        continue\n",
        "      #print(dimers[i][0],dimers[i][1],dimers[i][2])\n",
        "      spaces = \"\"\n",
        "      for a in range (len(rev) - len(dimers[i][0])):\n",
        "        spaces += \" \"\n",
        "      if (dimers[i][1] > 0):\n",
        "        spaces = spaces[0:len(spaces) - dimers[i][1]]\n",
        "      if (dimers[i][2] > 0):\n",
        "        spaces = spaces[0:len(spaces) - dimers[i][2]]\n",
        "      connect = spaces\n",
        "      for a in range (len(dimers[i][0])):\n",
        "        connect += \"|\"\n",
        "      print(spaces + forw)\n",
        "      print(connect)\n",
        "      print(reverse(rev))\n",
        "      bad = True\n",
        "    return bad"
      ],
      "metadata": {
        "id": "Kpy6-ryYkePz"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#primer dimer\n",
        "#input all the forward primers and then all the reverse primers\n",
        "#primers start with 5' and end with 3'\n",
        "#takes in a list of primers and returns the ones that may have primer dimers\n",
        "#with each other or by themselves\n",
        "fin = open ('sequence.txt', 'r')\n",
        "x = fin.readlines()\n",
        "for i in range (len(x)):\n",
        "  if (i != len(x) - 1):\n",
        "    x[i] = (x[i])[0:len(x[i]) - 1]\n",
        "for i in range (len(x)):\n",
        "  for a in range (i,len(x)):\n",
        "    if (i < len(x) / 2 and a < len(x) / 2):\n",
        "      print(str(i + 1) + \" forward and \" + str(a + 1) + \" forward:\")\n",
        "    elif (i < len(x) / 2 and a >= len(x) / 2):\n",
        "      print(str(i + 1) + \" forward and \" + str(int(a - len(x) / 2) + 1) + \" reverse:\")\n",
        "    else:\n",
        "      print(str(int(i - len(x) / 2) + 1) + \" reverse and \" + str(int(a - len(x) / 2) + 1) + \" reverse:\")\n",
        "    if (i == a):\n",
        "      continue\n",
        "    #  if (i < len(x) / 2):\n",
        "    #    same(x[i],True)\n",
        "    #  else:\n",
        "    #    same(complementary(x[i]),False)\n",
        "    else:\n",
        "      if (i < len(x) / 2 and a < len(x) / 2):\n",
        "        diff(x[i],flip(reverse(x[a])),1)\n",
        "      elif (i < len(x) / 2 or a < len(x) / 2):\n",
        "        if (diff(x[i],flip(reverse(x[a])),2)):\n",
        "          bad[a - int(len(x) / 2)].append(1)\n",
        "        else:\n",
        "          bad[a - int(len(x) / 2)].append(0)\n",
        "      else:\n",
        "        diff(x[i],flip(reverse(x[a])),1)\n",
        "#for i in range (8):\n",
        "  #print(bad[i])\n"
      ],
      "metadata": {
        "id": "akc2CkqTbwXv",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "efb5ee7f-89a6-466a-da69-1410498d9ad0"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1 forward and 1 forward:\n",
            "1 forward and 2 forward:\n",
            "1 forward and 3 forward:\n",
            "1 forward and 4 forward:\n",
            "1 forward and 5 forward:\n",
            "1 forward and 6 forward:\n",
            "1 forward and 1 reverse:\n",
            "1 forward and 2 reverse:\n",
            "1 forward and 3 reverse:\n",
            "1 forward and 4 reverse:\n",
            "1 forward and 5 reverse:\n",
            "1 forward and 6 reverse:\n",
            "2 forward and 2 forward:\n",
            "2 forward and 3 forward:\n",
            "2 forward and 4 forward:\n",
            "GGGAATATGCGGCGCGC\n",
            "             ||||\n",
            "             CGCGGGATGAGCGAGTGATAT\n",
            "2 forward and 5 forward:\n",
            "GGGAATATGCGGCGCGC\n",
            "             ||||\n",
            "             CGCGTTGGCGGTTTTATTTT\n",
            "2 forward and 6 forward:\n",
            "GGGAATATGCGGCGCGC\n",
            "             ||||\n",
            "             CGCGAGATGTTGGCGTTTG\n",
            "2 forward and 1 reverse:\n",
            "2 forward and 2 reverse:\n",
            "GGGAATATGCGGCGCGC\n",
            "            |||||\n",
            "            GCGCGCAGCAAATCCCAC\n",
            "2 forward and 3 reverse:\n",
            "2 forward and 4 reverse:\n",
            "GGGAATATGCGGCGCGC\n",
            "            ||||\n",
            "            GCGCAAATCAAAGCGCACAG\n",
            "2 forward and 5 reverse:\n",
            "GGGAATATGCGGCGCGC\n",
            "            ||||\n",
            "            GCGCAGCCAACCCAAAA\n",
            "2 forward and 6 reverse:\n",
            "3 forward and 3 forward:\n",
            "3 forward and 4 forward:\n",
            "3 forward and 5 forward:\n",
            "3 forward and 6 forward:\n",
            "3 forward and 1 reverse:\n",
            "3 forward and 2 reverse:\n",
            "3 forward and 3 reverse:\n",
            "TGGTTTGGTGTAGAAGCGTTC\n",
            "                |||\n",
            "                GCAGCGCAAATCGCAATCATC\n",
            "3 forward and 4 reverse:\n",
            "3 forward and 5 reverse:\n",
            "3 forward and 6 reverse:\n",
            "4 forward and 4 forward:\n",
            "4 forward and 5 forward:\n",
            "TATAGTGAGCGAGTAGGGCGC\n",
            "                 ||||\n",
            "                 CGCGTTGGCGGTTTTATTTT\n",
            "4 forward and 6 forward:\n",
            "TATAGTGAGCGAGTAGGGCGC\n",
            "                 ||||\n",
            "                 CGCGAGATGTTGGCGTTTG\n",
            "4 forward and 1 reverse:\n",
            "4 forward and 2 reverse:\n",
            "TATAGTGAGCGAGTAGGGCGC\n",
            "                 ||||\n",
            "                GCGCGCAGCAAATCCCAC\n",
            "4 forward and 3 reverse:\n",
            "4 forward and 4 reverse:\n",
            "TATAGTGAGCGAGTAGGGCGC\n",
            "                  |||\n",
            "                  GCGCAAATCAAAGCGCACAG\n",
            "4 forward and 5 reverse:\n",
            "TATAGTGAGCGAGTAGGGCGC\n",
            "                  |||\n",
            "                  GCGCAGCCAACCCAAAA\n",
            "4 forward and 6 reverse:\n",
            "5 forward and 5 forward:\n",
            "5 forward and 6 forward:\n",
            "TTTTATTTTGGCGGTTGCGC\n",
            "                ||||\n",
            "                CGCGAGATGTTGGCGTTTG\n",
            "5 forward and 1 reverse:\n",
            "5 forward and 2 reverse:\n",
            "TTTTATTTTGGCGGTTGCGC\n",
            "                ||||\n",
            "               GCGCGCAGCAAATCCCAC\n",
            "5 forward and 3 reverse:\n",
            "5 forward and 4 reverse:\n",
            "TTTTATTTTGGCGGTTGCGC\n",
            "                 |||\n",
            "                 GCGCAAATCAAAGCGCACAG\n",
            "5 forward and 5 reverse:\n",
            "TTTTATTTTGGCGGTTGCGC\n",
            "                 |||\n",
            "                 GCGCAGCCAACCCAAAA\n",
            "5 forward and 6 reverse:\n",
            "6 forward and 6 forward:\n",
            "6 forward and 1 reverse:\n",
            "6 forward and 2 reverse:\n",
            "GTTTGCGGTTGTAGAGCGC\n",
            "               ||||\n",
            "              GCGCGCAGCAAATCCCAC\n",
            "6 forward and 3 reverse:\n",
            "6 forward and 4 reverse:\n",
            "GTTTGCGGTTGTAGAGCGC\n",
            "                |||\n",
            "                GCGCAAATCAAAGCGCACAG\n",
            "6 forward and 5 reverse:\n",
            "GTTTGCGGTTGTAGAGCGC\n",
            "                |||\n",
            "                GCGCAGCCAACCCAAAA\n",
            "6 forward and 6 reverse:\n",
            "1 reverse and 1 reverse:\n",
            "1 reverse and 2 reverse:\n",
            "1 reverse and 3 reverse:\n",
            "1 reverse and 4 reverse:\n",
            "1 reverse and 5 reverse:\n",
            "1 reverse and 6 reverse:\n",
            "2 reverse and 2 reverse:\n",
            "2 reverse and 3 reverse:\n",
            "2 reverse and 4 reverse:\n",
            "CACCCTAAACGACGCGCG\n",
            "              ||||\n",
            "              GCGCAAATCAAAGCGCACAG\n",
            "2 reverse and 5 reverse:\n",
            "CACCCTAAACGACGCGCG\n",
            "              ||||\n",
            "              GCGCAGCCAACCCAAAA\n",
            "2 reverse and 6 reverse:\n",
            "CACCCTAAACGACGCGCG\n",
            "               |||\n",
            "             GCCGCCTCTAAGCGCTA\n",
            "3 reverse and 3 reverse:\n",
            "3 reverse and 4 reverse:\n",
            "3 reverse and 5 reverse:\n",
            "3 reverse and 6 reverse:\n",
            "4 reverse and 4 reverse:\n",
            "4 reverse and 5 reverse:\n",
            "GACACGCGAAACTAAACGCG\n",
            "                ||||\n",
            "                GCGCAGCCAACCCAAAA\n",
            "4 reverse and 6 reverse:\n",
            "GACACGCGAAACTAAACGCG\n",
            "                 |||\n",
            "               GCCGCCTCTAAGCGCTA\n",
            "5 reverse and 5 reverse:\n",
            "5 reverse and 6 reverse:\n",
            "AAAACCCAACCGACGCG\n",
            "              |||\n",
            "            GCCGCCTCTAAGCGCTA\n",
            "6 reverse and 6 reverse:\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#filter out pairs with bad orderings of g\n",
        "def gcheck(sseq):\n",
        "  g3 = []\n",
        "  g4 = False\n",
        "  for i in range (20,130):\n",
        "    if (sseq[i:i + 4] == 'gggg' or sseq[i:i + 4] == \"GGGG\"):\n",
        "      g4 = True\n",
        "      break\n",
        "    if (sseq[i:i + 3] == \"ggg\" or sseq[i:i + 3] == \"GGG\"):\n",
        "      g3.append(i)\n",
        "  for i in range (len(g3) - 1):\n",
        "    if (g3[i + 1] - g3[i] <= 10):\n",
        "      g4 = True\n",
        "      break;\n",
        "  if(g4):\n",
        "    return True\n",
        "  return False\n",
        "\n",
        "#input format for every set of primers and probe:\n",
        "#forward primer\n",
        "#probe\n",
        "#reverse primer\n",
        "#forward primer temperature\n",
        "#probe temperature\n",
        "#reverse primer temperature\n",
        "\n",
        "fin = open ('sequence.txt', 'r')\n",
        "fout = open ('newsequence.txt', 'w')\n",
        "x = fin.readlines()\n",
        "#post primer3 checking\n",
        "for a in range (len(x) / 6):\n",
        "  primerf = x[a * 6]\n",
        "  probe = x[a * 6 + 1]\n",
        "  primerr = x[a * 6 + 2]\n",
        "  if (gcheck(primerf)) :\n",
        "    continue\n",
        "  #Probe 5' end cannot be a G\n",
        "  if (probe[0] == 'G') :\n",
        "    continue\n",
        "  if (gcheck(primerr)) :\n",
        "    continue\n",
        "  primerftemp = int(x[a * 6 + 3])\n",
        "  probetemp = int(x[a * 6 + 4])\n",
        "  primerrtemp = int(x[a * 6 + 5])\n",
        "  #temperature of probe must be 6-8 C higher than primer\n",
        "  if (probetemp - primerftemp < 6 or probetemp - primerftemp > 8 or probetemp - primerrtemp < 6 or probetemp - primerrtemp > 8 ):\n",
        "    continue\n",
        "  print(primerf + \" \" + probe + \"  \" + primerr)\n"
      ],
      "metadata": {
        "id": "WYGDbybZWvNR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "k0qX3UJF-0VB",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7de2534d-330d-496f-ebcb-24bd23a2514d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CGCGCGCGGAGTTCGTTGAGATTTGAATTAATTTGGTTTAACGGTTTTTTTTAAATCGTTAGGAGTTTTTAATCGGCGGGATAGTAGGGCGCGGTGAGTTATCGTCGGTGATTAAGCGATTTTATTTTTTTTTTTCGGGTTTTTTTTTGTTATCGTCGTTTCGTAATTTTCGTCGTTCGAAGTTGGATTGAGTTCGTTAGGTTTTTCGATAGAATTTTTTTTTTTTTTAATATTTTTTCGTTAAGGTAAGTCGATGGATAGAGGCGCGGGTCGGAGTAGTTTTTTTTTTTAAGCGGGCGGCGCGCGAGGTTGCGGCGAGGTTTGAGTTTTGCGTTTTTGCGTTGTGCGCGTTTTTATTTCGCGTTTTAATTTTAGGCGTTTTTTGTTTTTTTTTTCGCGATTTTAGATTTGAGGGATTTTTTATTTTTTTTTTTTTTCGTTTTTTTGTTTGCGGGTTTTTTCGTTTGATCGTAGTTTCGAGATCGTCGCGTATTTTTTTTTACGTTTTTTTGGCGTGGTGTTATCGGATTTTTTTGGTTTAGTTTTAGGTGGATTTTTTTTTTATCGCGCGATTTCGTTTTTTTTAGTATTTTAGGGTGAGTTTAGTTTAGATTATTATTCGGAAAGTTTTTAAAAGTTTTAGTTTAGCGTTGAAGTAACGGGATTATGTTTAGTTTTAGGTTTCGGAGTAGGAAGGTTTGAGGGCGTTTTTATTTTATTCGTTTATTTTTTTCGTTTTTCGTTAGGTTTTTATTGGTTGGCGCGTTTCGCGGTTGGGATGGTAGTGGGAGGGGATTTTTTTTTTTAACGGGGTTATAAAAATAGCGTTTTCGGCGGGGTTTAGTTTTTTGTTATTTTCGTTTCGAGGTTTTCGCGTTAGAGACGTAGTCGCGTTTTTATTATTTATATTTATCGCGTTTTCGTTCGTTTTTTTTTCGGGAGTTAGTTCGCGTTATCGTCGTCGTTTAGGTTATCGTTATTTTTCGTAGTTATGTTTATTAGGTTCGTGTTTTCGTTTTTTTATCGTAGGATGTTCGGCGGTTCGGGTATCGCGAGTCGGTCGAGTTTTAGTCGGAGTTACGTGATTACGTTTATTCGTATTTATAGTTTGGGTAGCGCGTTGCGTTTTAGTATTAGTCGTAGTTTTTATGTTTCGTTTTCGGGCGGCGTGTATGTTACGCGTTTTTTTGTCGTGCGTTTGCGGAGTAGCGTGTTCGGGGTGCGGTTTTTGTAGGATTCGGTGGATTTTTCGTTGGTCGACGTTATTAATATCGAGTTTAAGAATATTCGTATTAACGAGAAGGTGGAGTTGTAGGAGTTGAATGATCGTTTCGTTAATTATATCGATAAGGTGCGTTTTTTGGAGTAGTAGAATAAGATTTTGTTGGTCGAGTTCGAGTAGTTTAAGGGTTAAGGTAAGTCGCGTTTGGGGGATTTTTACGAGGAGGAGATGCGGGAGTTGCGTCGGTAGGTGGATTAGTTAATTAACGATAAAGTTCGCGTCGAGGTGGAGCGCGATAATTTGGTCGAGGATATTATGCGTTTTCGGGAGAAGTAAGGTTGCGTTTATGTAAGTAGTTGGGTTTCGGGAGGGGGTTGGAGGGAGAGGGGAACGTTTTTTCGGTTTTCGCGAGAGTTGTTACGTTTTTGGGGATGTGGTCGGGGGGAGGTTTGTTAGGGAGATAGCGGAGAGCGGGGTTGTGGTTGTGGTGGCGTAGTTTCGTTTAGAATTTAGATTTTGTAGTTCGTATTTTTTTTTTTGTTTTTATATATTGTTTAAGGACGTTTCGTTTTAAGTTATAGATTTTTTAAAATTATTATTTTGTGTGTAGTTGAAGGTTTTTGGGTATAATGAGAGTTAGTTTTTTAAATTTTTAGAAAGTTTTTTGTTTTTTTTGGTAGGTTGTTAATTATCGGGCGGGAGAAGGAAGGAGGGGAAGGCGGTGGAGGGAGCGAGATAAAGGGATGGTTTTTCGGGGGCGGGGATGGCGGGGTTGTTTTGTAGGTTTGTGCGGTTATCGTGATTGTTTTTTTGCGCGGTGTTCGAAGTTTCGTTGAAATTTGTCGAGGGTAGTAGGTTTGAAAGTTGTAGGCGTTAGTTGCGCGGAGGTGGCGTAGTTGTTTTGGAGGCGTAGAGCGAATACGTGGTGTTTGGGTGTGGTCGTTTCGTTTTTGGCGGTTTTTTCG\n",
            "\n",
            "CGCGCGCGGAGTTCGTTGAGATTTGAATTAATTTGGTTTAACGGTTTTTTTTAAATCGTTAGGAGTTTTTAATCGGCGGGATAGTAGGGCGCGGTGAGTTATCGTCGGTGATTAAGCGATTTTATTTTTTTTTTTCGGGTTTTTTTTTGTTATCGTCGTTTCGTAATTTTCGTCGTTCGAAGTTGGATTGAGTTCGTTAGGTTTTTCGATAGAATTTTTTTTTTTTTTAATATTTTTTCGTTAAGGTAAGTCGATGGATAGAGGCGCGGGTCGGAGTAGTTTTTTTTTTTAAGCGGGCGGCGCGCGAGGTTGCGGCGAGGTTTGAGTTTTGCGTTTTTGCGTTGTGCGCGTTTTTATTTCGCGTTTTAATTTTAGGCGTTTTTTGTTTTTTTTTTCGCGATTTTAGATTTGAGGGATTTTTTATTTTTTTTTTTTTTCGTTTTTTTGTTTGCGGGTTTTTTCGTTTGATCGTAGTTTCGAGATCGTCGCGTATTTTTTTTTACGTTTTTTTGGCGTGGTGTTATCGGATTTTTTTGGTTTAGTTTTAGGTGGATTTTTTTTTTATCGCGCGATTTCGTTTTTTTTAGTATTTTAGGGTGAGTTTAGTTTAGATTATTATTCGGAAAGTTTTTAAAAGTTTTAGTTTAGCGTTGAAGTAACGGGATTATGTTTAGTTTTAGGTTTCGGAGTAGGAAGGTTTGAGGGCGTTTTTATTTTATTCGTTTATTTTTTTCGTTTTTCGTTAGGTTTTTATTGGTTGGCGCGTTTCGCGGTTGGGATGGTAGTGGGAGGGGATTTTTTTTTTTAACGGGGTTATAAAAATAGCGTTTTCGGCGGGGTTTAGTTTTTTGTTATTTTCGTTTCGAGGTTTTCGCGTTAGAGACGTAGTCGCGTTTTTATTATTTATATTTATCGCGTTTTCGTTCGTTTTTTTTTCGGGAGTTAGTTCGCGTTATCGTCGTCGTTTAGGTTATCGTTATTTTTCGTAGTTATGTTTATTAGGTTCGTGTTTTCGTTTTTTTATCGTAGGATGTTCGGCGGTTCGGGTATCGCGAGTCGGTCGAGTTTTAGTCGGAGTTACGTGATTACGTTTATTCGTATTTATAGTTTGGGTAGCGCGTTGCGTTTTAGTATTAGTCGTAGTTTTTATGTTTCGTTTTCGGGCGGCGTGTATGTTACGCGTTTTTTTGTCGTGCGTTTGCGGAGTAGCGTGTTCGGGGTGCGGTTTTTGTAGGATTCGGTGGATTTTTCGTTGGTCGACGTTATTAATATCGAGTTTAAGAATATTCGTATTAACGAGAAGGTGGAGTTGTAGGAGTTGAATGATCGTTTCGTTAATTATATCGATAAGGTGCGTTTTTTGGAGTAGTAGAATAAGATTTTGTTGGTCGAGTTCGAGTAGTTTAAGGGTTAAGGTAAGTCGCGTTTGGGGGATTTTTACGAGGAGGAGATGCGGGAGTTGCGTCGGTAGGTGGATTAGTTAATTAACGATAAAGTTCGCGTCGAGGTGGAGCGCGATAATTTGGTCGAGGATATTATGCGTTTTCGGGAGAAGTAAGGTTGCGTTTATGTAAGTAGTTGGGTTTCGGGAGGGGGTTGGAGGGAGAGGGGAACGTTTTTTCGGTTTTCGCGAGAGTTGTTACGTTTTTGGGGATGTGGTCGGGGGGAGGTTTGTTAGGGAGATAGCGGAGAGCGGGGTTGTGGTTGTGGTGGCGTAGTTTCGTTTAGAATTTAGATTTTGTAGTTCGTATTTTTTTTTTTGTTTTTATATATTGTTTAAGGACGTTTCGTTTTAAGTTATAGATTTTTTAAAATTATTATTTTGTGTGTAGTTGAAGGTTTTTGGGTATAATGAGAGTTAGTTTTTTAAATTTTTAGAAAGTTTTTTGTTTTTTTTGGTAGGTTGTTAATTATCGGGCGGGAGAAGGAAGGAGGGGAAGGCGGTGGAGGGAGCGAGATAAAGGGATGGTTTTTCGGGGGCGGGGATGGCGGGGTTGTTTTGTAGGTTTGTGCGGTTATCGTGATTGTTTTTTTGCGCGGTGTTCGAAGTTTCGTTGAAATTTGTCGAGGGTAGTAGGTTTGAAAGTTGTAGGCGTTAGTTGCGCGGAGGTGGCGTAGTTGTTTTGGAGGCGTAGAGCGAATACGTGGTGTTTGGGTGTGGTCGTTTCGTTTTTGGCGGTTTTTTCG\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "#Input: The sequence should be put in a file called sequence.txt.\n",
        "#Output: The changed sequence will be put in a file called newsequence.txt.\n",
        "#Find the data folder from the folder tab on the left tool bar.\n",
        "#Both files will have to be created outside of the sample data folder.\n",
        "#input is a sequence that contains cpg islands for a biomarker\n",
        "#output is a list of amplicons from the original sequence that could contain forward and reverse primers\n",
        "fin = open ('sequence.txt', 'r')\n",
        "fout = open ('newsequence.txt', 'w')\n",
        "x = fin.readlines()\n",
        "seq = \"\"\n",
        "for i in range (len(x)):\n",
        "  seq += x[i]\n",
        "newseq = \"\"\n",
        "used = False\n",
        "forward = -1\n",
        "seq2 = seq\n",
        "#seq = complementary(seq)\n",
        "#convert c to t in an unmethylated strand\n",
        "for i in range (len(seq)):\n",
        "  if (seq[i] == 'C' or seq[i] == 'c') :\n",
        "    if (i + 1 >= len(seq)):\n",
        "      if (seq[i] == 'C'):\n",
        "         newseq += 'T'\n",
        "      else:\n",
        "        newseq += 't'\n",
        "    elif (seq[i + 1] == 'G' or seq[i + 1] == 'g'):\n",
        "      newseq += seq[i]\n",
        "    else:\n",
        "      if (seq[i] == 'C'):\n",
        "        newseq += 'T'\n",
        "      else:\n",
        "        newseq += 't'\n",
        "  else:\n",
        "    if (seq[i] == 'T' or seq[i] == 't' or seq[i] == 'G' or seq[i] == 'g' or seq[i] == 'A' or seq[i] == 'a'):\n",
        "      newseq += seq[i]\n",
        "print(newseq)\n",
        "point = 0\n",
        "start = 0\n",
        "numsarr = []\n",
        "sseqarr = []\n",
        "first = True\n",
        "while True:\n",
        "  start = dense(newseq,start,first)\n",
        "  first = False\n",
        "  #filter out strands that dont have enough cpg pairs\n",
        "  if (start == -1):\n",
        "    if (len(numsarr) > 0):\n",
        "      print()\n",
        "      break\n",
        "    print(\"INVALID\")\n",
        "    break\n",
        "  sseq = newseq[start:start + 150]\n",
        "  #print(start)\n",
        "  #print(sseq)\n",
        "\n",
        "\n",
        "\n",
        "  #print(\"W\",sseq)\n",
        "  places = cgcgpairs(sseq)\n",
        "  #print(sseq)\n",
        "  large = False\n",
        "  for i in range (len(places)):\n",
        "    if (places[i] == '-'):\n",
        "      large = True\n",
        "      break\n",
        "  if (large):\n",
        "    start += 1\n",
        "    continue\n",
        "  #print(sseq)\n",
        "  #print(\"temp\",temp(sseq[splace:eplace + 2]))\n",
        "  #print(score)\n",
        "  #print(gcdensity(sseq[splace:eplace + 2]))\n",
        "  numsarr.append([score,start,splace,eplace,sseq,sseq[splace:eplace + 2],splace2,eplace2])\n",
        "  #numsarr.append([score,splace,eplace])\n",
        "  #print(numsarr[len(numsarr) - 1][0])\n",
        "  #sseqarr.append([sseq,sseq[splace:eplace + 2]])\n",
        "  #print(splace,eplace + 2)\n",
        "  #print(start)\n",
        "  start += 1\n",
        "  #break\n",
        "\n",
        "#do everything for the reverse strand\n",
        "print(newseq)\n",
        "point = 0\n",
        "start = 0\n",
        "sseq2arr = []\n",
        "first = True\n",
        "while True:\n",
        "  start = dense(newseq,start,first)\n",
        "  first = False\n",
        "  if (start == -1):\n",
        "    if (len(numsarr) > 0):\n",
        "      print()\n",
        "      break\n",
        "    print(\"INVALID\")\n",
        "    break\n",
        "  sseq2 = newseq[start:start + 150]\n",
        "  #print(start)\n",
        "  #print(sseq2)\n",
        "  g3 = []\n",
        "  g4 = False\n",
        "  #print(\"W\",sseq2,start)\n",
        "  for i in range (20,130):\n",
        "    if (sseq2[i:i + 4] == 'gggg' or sseq2[i:i + 4] == \"GGGG\"):\n",
        "      start += i + 1;\n",
        "      g4 = True\n",
        "      break\n",
        "    if (sseq2[i:i + 3] == \"ggg\" or sseq2[i:i + 3] == \"GGG\"):\n",
        "      g3.append(i)\n",
        "  for i in range (len(g3) - 1):\n",
        "    if (g3[i + 1] - g3[i] <= 10):\n",
        "      start += g3[i] + 1\n",
        "      g4 = True\n",
        "      break;\n",
        "  if(g4):\n",
        "    continue\n",
        "  #print(\"W\",sseq2)\n",
        "  places = cgcgpairs(sseq2)\n",
        "  #print(sseq2)\n",
        "  large = False\n",
        "  for i in range (len(places)):\n",
        "    if (places[i] == '-'):\n",
        "      large = True\n",
        "      break\n",
        "  if (large):\n",
        "    start += 1\n",
        "    continue\n",
        "  #print(sseq2)\n",
        "  #print(\"temp\",temp(sseq2[splace:eplace + 2]))\n",
        "  #print(score)\n",
        "  #print(gcdensity(sseq2[splace:eplace + 2]))\n",
        "  numsarr.append([score,start,splace,eplace,sseq2,sseq2[splace:eplace + 2],splace2,eplace2])\n",
        "  #numsarr.append([score,splace,eplace])\n",
        "  #print(numsarr[len(numsarr) - 1][0])\n",
        "  #sseq2arr.append([sseq2,sseq2[splace:eplace + 2]])\n",
        "  #print(splace,eplace + 2)\n",
        "  #print(start)\n",
        "  start += 1\n",
        "  #break\n",
        "\n",
        "#sort all primer candidates by score and output\n",
        "numsarr.sort()\n",
        "lastp = -2\n",
        "lastsub = \"-1\"\n",
        "for i in range(len(numsarr) - 1,-1,-1):\n",
        "  if (numsarr[i][1] == lastp - 1 and numsarr[i][5] == lastsub):\n",
        "      lastp -= 1\n",
        "      continue\n",
        "  elif (numsarr[i][1] == lastp and numsarr[i][5] == lastsub):\n",
        "      continue\n",
        "  else:\n",
        "    lastp = numsarr[i][1]\n",
        "    lastsub = numsarr[i][5]\n",
        "  fout.write(\"Place in original sequence \" + str(numsarr[i][1] + 1) + \" to \" + str(numsarr[i][1] + 150) + \"\\n\")\n",
        "  fout.write(\"Full sequence: \" + str(complementary(numsarr[i][4])) + \"\\n\")\n",
        "  #fout.write(\"Start and End points: \" + str(numsarr[i][6] + 1) + \" \" + str(numsarr[i][7] + 1) + \"\\n\")\n",
        "  #fout.write(\"Subsequence: \" + complementary(numsarr[i][5]) + \"\\n\")\n",
        "  #print(numsarr[i][6])\n",
        "  #print(\"Forward primer:\",numsarr[i][4][numsarr[i][6] - 19:numsarr[i][6] + 1])\n",
        "  #print(\"Reverse primer thing:\",complementary(numsarr[i][4][numsarr[i][7]:numsarr[i][7] + 20]))\n",
        "  print()\n",
        "#print(newseq[])\n",
        "fout.close()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "next steps:\n",
        "\n",
        "1.   download ucsc database of gene and parsing\n",
        "2.   score system remodeling\n",
        "3.   summarize rules\n",
        "4.   allow code to check top and bottom strand\n",
        "5.   make it into a usable multiplex assay\n",
        "\n"
      ],
      "metadata": {
        "id": "Yndz8McnwB_9"
      }
    }
  ]
}